### PR TITLE
Added Cardano offchain integration

### PR DESCRIPTION
**Pull request by Hathoriel**

### CODE REVIEW

File: src/transaction/ada.ts [hunk 1]
```diff
@@ -1,82 +1,218 @@
...
+const addOutput = (transactionBuilder: TransactionBuilder, address: string, amount: string) => {
+  transactionBuilder.add_output(TransactionOutput.new(
+    Address.from_bech32(address),
+    Value.new(BigNum.from_str(amount)),
+  ));
+}
+
+const addInput = (transactionBuilder: TransactionBuilder, privateKey: string, utxo: AdaUtxo) => {
+  const privateKeyCardano = Bip32PrivateKey.from_128_xprv(
+    Buffer.from(privateKey, 'hex'),
```
Comment (Hathoriel) [inline src/transaction/ada.ts:124]: @samuelsramko Why I need a here private key? In the case of BTC, this doesn't need the private key. With this setup, I am not able to create a transaction and sign it later. Please propose some workaround.

File: src/transaction/ada.ts [hunk 2]
```diff
@@ -1,82 +1,215 @@
...
+  const witnesses = TransactionWitnessSet.new();
+  witnesses.set_vkeys(vKeyWitnesses);
+  return Buffer.from(
+    Transaction.new(transactionBody, witnesses).to_bytes(),
+  ).toString('hex')
+};
+
+const addOutputs = (transactionBuilder: TransactionBuilder, tos: To[]) => {
+  let amount = 0
+  for (const to of tos) {
```
Comment (samuelsramko) [inline src/transaction/ada.ts:65]: do the whole calculation in BigNumber. let amount = new BigNumber(0)...

File: src/transaction/ada.ts [hunk 3]
```diff
@@ -1,82 +1,215 @@
...
+  throw new Error('Field fromAddress or fromUTXO is not filled.')
+}
+
+const addAddressInputs = async (transactionBuilder: TransactionBuilder, fromAddresses: FromAddress[]) => {
+  let amount = 0
+  const privateKeysToSign: string[] = [];
+  for (const fromAddress of fromAddresses) {
+    const { address, privateKey } = fromAddress
+    const utxos: AdaUtxo[] = await adaGetUtxos(address)
+    for (const utxo of utxos) {
```
Comment (samuelsramko) [inline src/transaction/ada.ts:90]: use BigNumber()

File: src/transaction/ada.ts [hunk 4]
```diff
@@ -1,82 +1,215 @@
...
+      amount += parseInt(utxo.value);
+      addInput(transactionBuilder, privateKey, utxo, address)
+      privateKeysToSign.push(fromAddress.signatureId || fromAddress.privateKey)
+    }
+  }
+  return { amount, privateKeysToSign }
+}
+
+const addUtxoInputs = async (transactionBuilder: TransactionBuilder, fromUTXOs: FromUTXO[]) => {
+  let amount = 0
```
Comment (samuelsramko) [inline src/transaction/ada.ts:100]: use BigNumber

**Review by samuelsramko — CHANGES_REQUESTED** [review]

**Review by samuelsramko — APPROVED** [review]

**Pull request merged**